modules/new_serverpod_auth/serverpod_auth_email_account/serverpod_auth_email_account_server/lib/src/business/email_accounts.dart

```
@@ -0,0 +1,314 @@
+        .where((final r) => r.expiration.isAfter(DateTime.now()))
+        .firstOrNull;
+
+    if (request == null) {
+      return null;
+    }
+
+    return (request.id!, request.email);
+  }
+
+  /// Finalize the email authentication creation.
+  ///
+  /// Returns the `ID` of the new email authentication, and the email address used during registration.
+  static Future<(UuidValue emailAuthenticationId, String emailAddress)>
+      createAccount(
+    final Session session, {
+    required final String verificationCode,
+
+    /// Authentication user ID this account should be linked up with
+    required final UuidValue userId,
+  }) async {
+    final request = (await EmailAccountRequest.db.find(session,
+            where: (final t) => t.verificationCode.equals(verificationCode)))
+        .where((final r) => r.expiration.isAfter(DateTime.now()))
+        .singleOrNull;
+
+    if (request == null) {
+      throw Exception('Email account request not found');
+    }
+
+    await EmailAccountRequest.db.deleteRow(session, request);
+
+    final account = await EmailAccount.db.insertRow(
+      session,
+      EmailAccount(
+        userId: userId,
+        created: DateTime.now(),
+        email: request.email,
+        passwordHash: PasswordHash.createHash(
+          email: request.email,
+          password: request.passwordHash,
+        ),
+      ),
+    );
+
+    return (account.id!, request.email);
+  }
+
+  /// Sends out a password reset email for the given account.
+  static Future<void> requestPasswordReset(
+    final Session session, {
+    required String email,
+  }) async {
+    email = email.trim().toLowerCase();
+
+    final resetToken = Random.secure().nextString(length: 20);
+
+    final account = await EmailAccount.db.findFirstRow(
+      session,
+      where: (final t) => t.email.equals(email),
+    );
+
+    if (account == null) {
+      throw Exception('No account found for email.');
```

Should we instead silently return here and in `requestAccount`, so that the outside shows a text: If an account exist, you'll get an email now. (And for `requestAccount`: If the account did not exist already, you'll get an email now.)

Because currently this could be used to check the existence of accounts.

But this might also be a higher-level concern, because at least for `requestAccount` we need to know whether a request was created and potentially store additional data with it.

But maybe an explicit error would also explain/document this better for the developer.

---

This is a security consideration. The requestPasswordReset endpoint should not leak whether an account exists or not so I don't think we should throw, at least not from an endpoint perspective.

Do we trust our users to always catch this exception when they override the usage of this method? Otherwise we might add this as a configurable callback instead where they can choose what should happen?

---

Decision:

- In all these methods around "user not found" do not throw (which could inadvertently leak whether or not the account exists through the endpoint implementation)
- Instead return the status through an enum return value. The endpoint might just ignore it and always state that a mail might have been sent, whereas internal code can act upon the result.


```
Future<EmailResetStatus> resetPassword(…);
```

---

Updated the implementation to return status enums and not throw in "email unknown" or "email already registered" case.

That way the caller can not accidentally leak that information.